nucypher/utilities/versioning.py

```diff
@@ -0,0 +1,160 @@
+
+    @classmethod
+    def from_bytes(cls, data: bytes):
+        """"Public deserialization API"""
+        if len(data) < cls._HEADER_SIZE:
+            raise ValueError(f'Invalid bytes for {cls.__name__}.')
+        brand, version, payload = cls._parse(data)
+        major, minor = version
+        latest_major_version = cls._version()[0]
+        handlers = cls._deserializers()
+        compatible = major == latest_major_version
+        if not compatible:
+            message = f'Incompatible versioned bytes for {cls.__name__}. ' \
+                      f'Compatible version is {cls.version_string()}; Got {major}.{minor}.'
+            raise cls.IncompatibleVersion(message)
+        return handlers[version](payload)
+
+    @classmethod
+    def _parse(cls, data: bytes) -> Tuple[bytes, Tuple[int, int], bytes]:
+        brand = cls._parse_brand(data)
+        version = cls._parse_version(data)
+        payload = cls._parse_payload(data)
+        return brand, version, payload
+
+    @classmethod
+    def _parse_brand(cls, data: bytes) -> bytes:
+        brand = data[:cls._BRAND_SIZE]
+        if brand != cls._brand():
+            error = f"Incorrect brand. Expected {cls._brand()}, Got {brand}."
+            if not brand.isalpha():
+                # unversioned entities for older versions will most likely land here.
+                error = f"Incompatible bytes for {cls.__name__}."
+            raise cls.InvalidHeader(error)
+        return brand
+
+    @classmethod
+    def _parse_version(cls, data: bytes) -> Tuple[int, int]:
+        version_data = data[cls._BRAND_SIZE:cls._HEADER_SIZE]
+        major, minor = version_data[:cls._PART_SIZE], version_data[cls._PART_SIZE:]
+        major, minor = int.from_bytes(major, 'big'), int.from_bytes(minor, 'big')
+        version = major, minor
+        known_version = version in cls._deserializers()
+        if not known_version:
+            # TODO: Separately handle (potentially) future versions here?
+            available_versions = ", ".join(f'{str(v[0])}.{str(v[1])}' for v in sorted(cls._deserializers()))
+            error = f'Incorrect or unknown version "{major}.{minor}". Available versions for {cls.__name__} {available_versions}'
+            raise cls.IncompatibleVersion(error)
+        return version
+
+    @classmethod
+    def _parse_payload(cls, data: bytes) -> bytes:
+        payload = data[cls._HEADER_SIZE:]
+        if len(payload) == 0:
+            raise ValueError(f'No content to deserialize {cls.__name__}.')
+        return payload
+
+    @classmethod
+    def _deserializers(cls) -> Dict[Tuple[int, int], Callable]:
+        """Return a dict of all known deserialization handlers for this class keyed by version"""
+        return {cls._version(): cls._from_bytes_current, **cls._old_version_handlers()}
+
+
+# Collects the brands of every serializable entity
+SERIALIZABLE_ENTITIES = {v.__class__.__name__: v._brand() for v in Versioned.__subclasses__()}
```

Comment:
Is this used somewhere?

Reply:
✔️ Commented it out and left an explainer.